src/history/recorder.js

```diff
@@ -41,34 +28,59 @@ export default class Recorder {
       await fs.mkdir(directory, { recursive: true });
     }
 
-    const filePath = `${directory}/${documentType}.${this.fileExtension}`;
+    const filePath = this.getPathFor(serviceId, documentType);
 
     await fs.writeFile(filePath, content);
 
     return filePath;
   }
 
   async commit(filePath, message) {
-    if (!await this.git.hasChanges(filePath)) {
-      return;
+    try {
+      if (!await this.git.hasChanges(filePath)) {
+        return;
+      }
+
+      await this.git.add(filePath);
+      return await this.git.commit(filePath, message);
+    } catch (error) {
+      throw new Error(`Could not commit ${filePath} with message "${message}" due to error: "${error}"`);
     }
+  }
 
-    return new Promise((resolve, reject) => {
-      this.commitQueue.push({ filePath, message, resolve, reject });
-    });
+  async publish() {
+    return this.git.pushChanges();
   }
 
-  async _commit({ filePath, message, resolve }) {
-    await this.git.add(filePath);
-    resolve(await this.git.commit(filePath, message));
+  async getLatestRecord(serviceId, documentType) {
+    const filePath = this.getPathFor(serviceId, documentType);
+    const commits = await this._getCommits(filePath);
+    const result = { id: null, path: filePath };
+
+    if (commits.length) {
+      result.id = commits[0].hash;
+    }
+
+    return result;
   }
 
-  async publish() {
-    return this.git.pushChanges();
+  getPathFor(serviceId, documentType) {
+    return `${this.path}/${serviceId}/${documentType}.${this.fileExtension}`;
+  }
+
+  async isTracked(serviceId, documentType) {
+    const filePath = this.getPathFor(serviceId, documentType);
+
+    if (!await fileExists(filePath)) {
+      return false;
+    }
+
+    const isNew = await this.git.isNew(filePath);
+    return !isNew;
```

Comment:
This method is called on each record. It is immensely more likely that the file is already tracked than it is new, and so that the initial `fileExists` call is useless while still incurring a filesystem access. We could thus cut the cost of this call by two in the immense majority of cases by inverting the tests: can we call `git.isNew` with the expectation that the file exists? 🙂